src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -127,10 +128,14 @@ public static IEnumerable<DType> GetPrimitiveTypes()
         // This is only used by attachment type. DocumentDataType is used to workaround the issue of having cycles in struct type.
         protected readonly DType _attachmentType;
 
-        internal HashSet<IExternalTabularDataSource> AssociatedDataSources { get; private set; }
+        internal HashSet<IExternalTabularDataSource> AssociatedDataSources { get; }
         internal IExternalOptionSet<int> OptionSetInfo { get; }
         internal IExternalViewInfo ViewInfo { get; }
 
+        // Eventually, all display names should come from this centralized source
+        // We should not be using individual DataSource/OptionSet/View references
+        internal HashSet<BidirectionalDictionary<string, string>> DisplayNameProviders { get; }
```

Comment:
I think this might be worth abstracting to a little interface, so that lookup handling can be isolated into a combo of e.g. `StaticDisplayNameProvider`s and `CompositeDisplayNameProvider`s.

Reply:
This can also help to manage immutability/copying